src/manager.rs

```diff
@@ -407,6 +354,128 @@ impl ManagerService for Arc<Manager> {
+            ))
+        })?;
+
+    let step = participants[replica_rank].step;
+
+    // Compute the details for workers at max step.
+    let max_step = participants.iter().map(|p| p.step).max().unwrap();
+    let max_participants: Vec<&QuorumMember> =
+        participants.iter().filter(|p| p.step == max_step).collect();
+    let max_rank = max_participants.iter().enumerate().find_map(|(i, p)| {
+        if p.replica_id == replica_id {
+            Some(i as i64)
+        } else {
+            None
+        }
+    });
+
+    // The primary TCPStore to use for this rank.
+    let primary_rank = rank as usize % max_participants.len();
+    let primary = max_participants[primary_rank];
+
+    // Compute recovery assignments
+
+    // Nodes are recovering if:
+    // 1. not at the max step
+    // 2. max_step == 0 and not the primary replica
+    let all_recovering_ranks: Vec<usize> = participants
+        .iter()
+        .enumerate()
+        .filter_map(|(i, p)| {
+            if p.step != max_step || max_step == 0 && primary.replica_id != p.replica_id {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .collect();
+    let all_recovering_ranks_set = all_recovering_ranks.iter().collect::<HashSet<_>>();
+    let serving_ranks: Vec<usize> = participants
+        .iter()
+        .enumerate()
+        .filter_map(|(i, _p)| {
+            if !all_recovering_ranks_set.contains(&i) {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // This is a map of rank to the ranks that are recovering from that node.
+    let mut recovery_assignments: HashMap<usize, Vec<i64>> = HashMap::new();
+    // The rank of the node that this rank is recovering from.
+    let mut recover_rank: Option<i64> = None;
+    for (i, recovering_rank) in all_recovering_ranks.iter().enumerate() {
+        let serving_idx = (i + rank as usize) % serving_ranks.len();
+        let recovering_recover_rank = serving_ranks[serving_idx];
+        if !recovery_assignments.contains_key(&recovering_recover_rank) {
+            recovery_assignments.insert(recovering_recover_rank, Vec::new());
+        }
+        recovery_assignments
+            .get_mut(&recovering_recover_rank)
+            .unwrap()
+            .push(*recovering_rank as i64);
+        if *recovering_rank == replica_rank {
+            recover_rank = Some(recovering_recover_rank as i64);
+        }
+    }
+
+    info!("serving_ranks={:?}", serving_ranks);
+
+    let heal = recover_rank.is_some();
+    if heal {
+        info!(
+            "healing is required step={}, max_step={}, recover_rank={}",
+            step,
+            max_step,
+            recover_rank.unwrap()
+        );
+    }
+
+    Ok(ManagerQuorumResponse {
+        quorum_id: quorum.quorum_id,
+        // address is used for looking up the checkpoint server address.
+        recover_manager_address: primary.address.clone(),
```

Comment:
how/why do we use the `recover_manager_address` if the checkpoint loading is done round robin?